File: partystatus.py
"""What each family member is doing, said beside their name in the game.

infra#3334. Five clients each stream one member, and on every one of them the
party frames name the other four and say nothing about what they are doing.
The Family tab and the Watch wall answer that on a web page; nobody watching a
stream is looking at a web page.

TWO SOURCES, AND THE SPLIT IS THE WHOLE DESIGN. Five of these states are
already sitting in the client: UnitIsConnected, UnitIsDeadOrGhost,
UnitIsVisible (the client is not tracking the unit at all, so it is a zone away
or far outside sight), UnitAffectingCombat, and UnitInRange - the ~40 yard
check Blizzard_RaidUI uses. They are observed rather than reported, true within
a frame of being true, and they keep working when every process behind this one
is down. PRECEDENCE below is that list, in the order it is asked.

The rest is only knowable here. Whether a character is ON TASK, walking an
errand, inside a run, or being driven by nothing lives in `overseer_roster` and
`overseer_dungeon_run`, and no client can see a database. That half is composed
here into one short line and pushed to the clients; the addon splits it and
renders it.

THE MODULE OWNS THE WORDS, THE ADDON OWNS THE COLOURS. Same seam as
watchwall.py (infra#2597): a module that returns "#ff4d2e" has decided what the
palette is, and a Lua file that decides what "on task" means has put judgement
somewhere no test can reach.

WHAT THE ADDON MUST MIRROR, and why that is safe. The precedence between the
five local facts has to run inside the client, because the facts do, and there
is no way to move it here. So it is written here as PRECEDENCE, written there
as LOCAL_RULES, and tests/test_partystatus.py reads the Lua as text and fails
if the two orders differ - the guard test_jobs.py puts on mod_overseer.cpp and
test_frames.py puts on map_server.py.

DEGRADING IS THE NORMAL CASE, not the failure case. Nothing runs the sender by
default, so an addon installed on its own shows the five local states and says
nothing else. It must never fill that silence with a guess: a member standing
quietly with the party and a member nobody is driving look identical from
inside the client, and only one of them is a problem.

THREE THINGS THIS DELIBERATELY DOES NOT SAY:

1. Health. The party frame draws a health bar six pixels away, and a second
   opinion about it would disagree the moment one of them rounded.
2. The dungeon PHASE. DungeonRunPhase (Gathering, Barrier, Enter, Clearing,
   ...) lives in world-thread memory in mod_overseer.cpp, is not persisted and
   is not queryable. `overseer_dungeon_run.state` is active or ended, so that
   is what DUNGEON means here and no more.
3. Held on the spot. The module logs a refusal when it will not walk a
   character off a drop, guarded by an in-process bool no table records. There
   is nothing to read, so there is no code for it: a state that can never
   arrive teaches a viewer to expect a warning that will not come.

HOW IT IS CARRIED. `overseer_command` already has `kind='chat'` and
`channel='party'`, and mod_overseer.cpp's DoChat broadcasts that to the group
as a packet it builds itself, so `build_push` emits ready-made rows and nothing
new is needed server-side. Two properties make party chat survivable for
machine traffic: the payload is TAB separated, which relay.is_addon_traffic
already recognises so it never reaches Discord, and the addon filters its own
lines out of the chat frames so nothing appears on the stream. The better route
is the same packet sent with LANG_ADDON, which no chat frame renders at all -
a few lines in DoChat, in quadseven/mod-overseer, and not this change. The
addon accepts both.

PURE MODULE: rows in, lines out. No MySQL, no client, and no clock of its own
unless one is not handed to it.
"""
from __future__ import annotations

from datetime import datetime

import agenda
import jobs

# --- the vocabulary ----------------------------------------------------------

# What the client can see for itself, in the order it is asked. First true one
# wins.
#
# AWAY OUTRANKS FIGHTING, the one place this disagrees with watchwall.py. That
# is a health-and-danger view, so a fight is the most urgent thing on it. This
# is a has-anyone-drifted-off view: a member the client cannot even see is the
# exact failure it exists to catch, and a drifted member is nearly always
# fighting something, so combat first would hide it behind the most ordinary
# word in the game. APART sits BELOW fighting for the opposite reason - forty
# yards during a fight is a caster standing where a caster stands, and a label
# that lit up on every pull would be ignored inside a day.
OFFLINE = "offline"
DEAD = "dead"
AWAY = "away"
FIGHTING = "fighting"
APART = "apart"
PRECEDENCE = (OFFLINE, DEAD, AWAY, FIGHTING, APART)

# The words the addon shows for those five. Here rather than in the Lua so
# every sentence a viewer reads is written in one file, and so the suite can
# hold the Lua to them.
LOCAL_LABELS = {
    OFFLINE: "offline",
    DEAD: "dead",
    AWAY: "away",
    FIGHTING: "fighting",
    APART: "apart",
}

# What only this side knows. TASK is the operator's "on task": something is
# driving this character and it is not one of the more specific cases below.
TASK = "task"
TRAVEL = "travel"
DUNGEON = "dungeon"
STALLED = "stalled"
IDLE = "idle"
PUSHED = (TASK, TRAVEL, DUNGEON, STALLED, IDLE)

CODES = PRECEDENCE + PUSHED

# --- the wire ----------------------------------------------------------------

# The prefix, which is also the addon-message prefix, matching the "DC"
# convention mod-dungeon-clear already uses in this worldserver. Nothing in
# OllamaChat.BlacklistCommands starts with it, and that list is prefix-matched
# on a word boundary, so it cannot be caught by accident.
PREFIX = "OVSR"

# Carried in the line rather than assumed. The addon and the server are
# deployed by completely different means - a file copied onto a Windows box
# against a container roll - so they will be out of step routinely, and an
# addon reading a version it does not know must fall back to the local states.
VERSION = "1"

# TAB, load-bearing twice over. relay.is_addon_traffic already treats any C0
# byte as machine traffic and keeps the row out of Discord and the thought
# store, so this needs no change there; and the client will not transmit a tab
# in chat, so no player can forge one of these lines by typing.
SEP = "\t"

# Two caps for one column: overseer_command.command is VARCHAR(255), and MySQL
# truncates rather than refusing. A line trimmed by the database arrives with
# its last field cut in half.
MAX_LABEL_CHARS = 14
MAX_LINE_CHARS = 240

# How long a pushed state may go unrefreshed before the addon drops back to the
# local states alone.
#
# NINETY SECONDS, from the sender's poll budget rather than taste.
# mod_overseer.cpp drains overseer_command every two seconds, so a sender
# running every thirty gets three attempts inside this window. Shorter and an
# ordinary worldserver hiccup blanks every label; longer and a dead sender
# leaves "on task" under a character who stopped an hour ago.
PUSH_STALE_SECONDS = 90


def _short(text) -> str:
    """A label trimmed to what fits beside a name, with no ellipsis.

    An ellipsis costs three of the fourteen characters to say something was
    cut: "guild busine" is read correctly and "guild bus..." is not read at all.
    """
    return str(text or "").strip()[:MAX_LABEL_CHARS]


def _clean(text) -> str:
    """A field with the separator taken out of it.

    A label carrying a tab would silently become two fields and shift every
    later member onto the wrong name. Replaced rather than refused: a label is
    cosmetic and a shifted line is a lie.
    """
    return str(text or "").replace(SEP, " ").replace("\n", " ").strip()


def line(members: list[dict]) -> str:
    """One line carrying the whole family's pushed state.

    ONE LINE FOR FIVE MEMBERS, not five lines. Five rows is five packets, five
    chat-frame filters and five chances for a partial update to render a family
    half in one state and half in another. At roughly twenty-five bytes a
    member the whole roster fits inside one 255-byte command with room to
    spare, and a viewer either sees the new picture or the old one.

    Members that would push the line past MAX_LINE_CHARS are dropped from the
    end rather than shortened further. A truncated line is not parseable; a
    short one is.
    """
    out = [PREFIX, VERSION]
    for member in members:
        fields = [_clean(member.get("name")), _clean(member.get("code")),
                  _short(_clean(member.get("label")))]
        if len(SEP.join(out + fields)) > MAX_LINE_CHARS:
            break
        out.extend(fields)
    return SEP.join(out)


def parse(text) -> list[dict] | None:
    """A line back into members, or None if it is not one of ours.

    None and [] are different answers and both are real. None means "not our
    traffic" - somebody else's addon, a person talking, a version this build
    does not know - and the caller must leave whatever it is showing alone. []
    means "ours, and it says nobody", which should blank the labels.

    A trailing partial triple is dropped rather than guessed at. It only
    happens when something truncated the line, and the fields that did arrive
    are still correctly paired with their names.
    """
    raw = str(text or "")
    if SEP not in raw:
        return None
    fields = raw.split(SEP)
    if len(fields) < 2 or fields[0] != PREFIX or fields[1] != VERSION:
        return None
    body = fields[2:]
    members = []
    for i in range(0, len(body) - 2, 3):
        name, code, label = body[i], body[i + 1], body[i + 2]
        if not name or code not in CODES:
            continue
        members.append({"name": name, "code": code, "label": label})
    return members


# --- the decision ------------------------------------------------------------

def decide(facts: dict, pushed: dict | None = None) -> dict:
    """What one character's label says, given what is known about them.

    THIS IS THE FUNCTION THE ADDON MIRRORS, and the only one it does. `facts`
    is what the client observed; `pushed` is the last line that arrived for
    this character, or None when none has or the last one went stale.

    Local first, always. A pushed state is at best seconds old and was composed
    from a snapshot; "he is dead" observed in this frame outranks "he is
    questing" recorded a minute ago, and a character who is offline is not
    doing any of the things the roster says.

    An unknown fact is a fact that is NOT true. The defaults are all the
    reassuring direction on purpose: a caller that hands in nothing gets the
    pushed state or silence, never five alarms about a member it failed to ask
    about.
    """
    if not facts.get("connected", True):
        return {"code": OFFLINE, "label": LOCAL_LABELS[OFFLINE]}
    if facts.get("dead"):
        return {"code": DEAD, "label": LOCAL_LABELS[DEAD]}
    if not facts.get("visible", True):
        return {"code": AWAY, "label": LOCAL_LABELS[AWAY]}
    if facts.get("combat"):
        return {"code": FIGHTING, "label": LOCAL_LABELS[FIGHTING]}
    if not facts.get("in_range", True):
        return {"code": APART, "label": LOCAL_LABELS[APART]}
    if pushed and pushed.get("code") in PUSHED:
        return {"code": pushed["code"], "label": _short(pushed.get("label"))}
    return {"code": "", "label": ""}


# --- composing the pushed half -----------------------------------------------

def _job_label(job: str) -> str:
    """The job column said in the space beside a name.

    jobs.MODES are already phrases a person says out loud, so they are printed
    as stored rather than translated into a second vocabulary.

    THE QUALIFIED DUNGEON JOB IS STRIPPED. mod_overseer.cpp's IsDungeonJob
    accepts `dungeon:deadmines` as well as bare `dungeon`, and jobs.py has no
    vocabulary for the qualified form at all. Printing it raw would spend the
    whole label on a keyword nobody needs beside a name.
    """
    text = str(job or "").strip() or jobs.DEFAULT
    return _short(text.split(":", 1)[0])


def _in_run(run: dict | None, roster: list[str]) -> set:
    """Who the active run counts as being in it.

    `members` is stamped at the moment the party is all inside, so it is empty
    for the whole approach - the reset, the walk to the staging point, the
    gather at the door. An empty column during an active run means "not stamped
    yet", not "nobody", and reading it the other way would say IDLE at five
    characters walking to a dungeon together.
    """
    if run is None:
        return set()
    stamped = [n.strip() for n in str(run.get("members") or "").split(",")
               if n.strip()]
    return set(stamped) if stamped else set(roster)


def build_push(roster_rows: list[dict], run_rows: list[dict],
               event_rows: list[dict], now: datetime | None = None) -> dict:
    """The pushed half of every label, plus the command rows that carry it.

    Reads what the agenda banner already reads, through agenda's own functions,
    so the party frames and the web page cannot come to different conclusions
    about the same family. Nothing new is queried and no table is added.

    THE ORDER OF THESE BRANCHES IS THE DESIGN:

    1. Not on the enabled roster - nothing is driving this character, and a
       disabled row is not a member for this purpose. IDLE, which is the
       operator's "not doing anything".
    2. The family is stalled  - nothing has happened anywhere in twenty
       minutes. It outranks every busy-looking state below BECAUSE they look
       busy: a run whose row still reads active and a job that is still set are
       exactly what a stuck family looks like from the tables, and
       quadseven/mod-overseer#171 is that state going unnoticed.
    3. Inside a run           - the most specific and most time-bound thing the
       family can be doing.
    4. Walking an errand      - a travel aim is per-character by design, and
       the column IS the status: the module clears it on arrival and on giving
       up, so non-empty means still walking.
    5. On task                - the ordinary case, named by the job.

    Stalled is family-wide and is therefore said about everyone, including a
    character standing somewhere blameless. The stall is a fact about the
    party, the operator is looking at five streams, and telling only one of
    them would make it look like one member's problem.
    """
    now = now or datetime.utcnow()
    orders = agenda.standing_orders(roster_rows)
    roster = orders["roster"]
    run = agenda.active_run(run_rows)
    is_stalled = agenda.stalled(agenda.last_movement(event_rows), now)
    inside = _in_run(run, roster)
    aims = {t["name"]: str(t.get("target") or "").strip()
            for t in orders["travel"]}
    job = _job_label(orders["job"])

    members = []
    for name in roster:
        if is_stalled:
            code, label = STALLED, "stalled"
        elif name in inside:
            code, label = DUNGEON, "dungeon"
        elif aims.get(name):
            code, label = TRAVEL, "walking"
        else:
            code, label = TASK, job
        members.append({"name": name, "code": code, "label": _short(label)})

    text = line(members)
    return {
        "generated_at": now.strftime("%Y-%m-%dT%H:%M:%S"),
        "stale_after_seconds": PUSH_STALE_SECONDS,
        "members": members,
        "line": text,
        "commands": commands(text, orders["leader"]),
    }


def commands(text: str, speaker: str | None) -> list[dict]:
    """The overseer_command rows that put one line in front of five clients.

    Party chat, spoken by the leader, because DoChat builds that packet itself
    and broadcasts it to the whole group: one row reaches all five sessions and
    none of them has to be picked as a special case.

    NO SPEAKER, NO ROWS. DoChat answers "not in a group" and marks the row an
    error, so enqueueing anyway would fill the queue with failures rather than
    reporting the honest fact that there is nobody to speak.

    Returned as rows rather than inserted: this module holds no connection and
    never will, and whoever inserts them owns the schedule and the retry.
    """
    if not speaker or not text:
        return []
    return [{
        "target_name": speaker,
        "command": text,
        "kind": "chat",
        "channel": "party",
    }]

File: wow-addons/PartyStatus/PartyStatus.lua
-- Party Status (infra#3334)
--
-- WHAT THIS IS. One small label beside each family member's name saying what
-- they are doing. Five clients each stream one of them, and on every one of
-- those streams the party frames name the other four and say nothing else.
--
-- WHERE THE JUDGEMENT LIVES, AND IT IS NOT HERE. Every state, every word a
-- viewer reads, and the order the local facts are asked in are decided in
-- production/scripts/wow-overseer/partystatus.py and covered by its stdlib
-- suite. This file gathers facts, applies that order, and paints. The only
-- thing it decides on its own is the palette, which is a view's business.
--
-- LOCAL_RULES BELOW MUST STAY IN THE SAME ORDER AS partystatus.PRECEDENCE.
-- It cannot be moved to Python - the facts only exist inside a running client
-- - so tests/test_partystatus.py reads this file as text and fails if the two
-- lists differ. Reorder one and the suite tells you about the other.
--
-- WHERE IT SITS, AND WHY IT COVERS NOTHING. Measured against 3.3.5a's own
-- PartyFrameTemplates.xml rather than by eye. A party member frame is 128x53
-- from its TOPLEFT, and every pixel of that is already spoken for:
--
--     portrait      (7,-6) 37x37          name       bottom-left at (50,-10)
--     health bar    (47,-12) 70x8         mana bar   (47,-21) 70x8
--     debuffs       (48,-32) 4 x 15x15    role icon  (7,-33) 19x19
--     pet frame     64x26, hanging BELOW the frame, and it MOVES:
--                   PartyMemberFrame_UpdatePet re-anchors it on every update,
--                   to (23,-43) when that member has a pet and (23,-27) when
--                   it does not, overriding the (20,-47) in the XML
--
-- The next party frame is anchored to that pet frame, so the pitch is 79 units
-- with a pet and 63 without - NOT the 83 the XML anchor alone suggests, which
-- is what reading PartyFrameTemplates.xml without reading PartyMemberFrame.lua
-- gets you. None of the five classes here carries a pet at this level, so 63 is
-- the live case; a PrintWindow capture of a running client measures the pitch
-- at 73 screen pixels against 1.111 pixels per unit, derived from the exactly
-- 9-unit gap between each frame's own health and mana bars. Either way there is
-- no free space inside the frame and none underneath it, and the tighter the
-- pitch the truer that is. The label
-- goes OUTSIDE the right edge instead, at the height of the name, which is
-- what "next to their name" means here. Two things live near that edge and
-- neither is reached: PartyMemberBackground, the optional panel behind the
-- whole party, is 134 wide from x=-5 and so ends at x=129; the voice-chat
-- speaker icon is at x=123 and is hidden unless voice chat is on, which it is
-- not on this realm. CLEAR_X = 6 puts the label's left edge at x=134.
--
-- THE PLAYER'S OWN FRAME IS ANCHORED DIFFERENTLY, because the same rule does
-- not fit there. PlayerFrame ends at screen x=213 and TargetFrame begins at
-- x=250, so the gap to its right is 37 pixels - narrower than the words. The
-- free space near PlayerFrame is underneath it: the pet frame ends at y=-117
-- and the party block does not begin until y=-154, so the label goes at
-- y=-120, left-aligned with the party column. On a stream of one character
-- that label is the one the viewer most wants, so leaving the player out was
-- not an option.
--
-- IT DEGRADES TO SILENCE, NOT TO A GUESS. Nothing carries the pushed line
-- today. With none arriving, the five locally observed states still work and
-- everything else shows nothing at all, because a member standing quietly with
-- the party and a member nobody is driving are indistinguishable from inside
-- the client, and only one of them is a problem.

local PREFIX = "OVSR"
local VERSION = "1"
local SEP = "\t"

-- Mirrors partystatus.PUSH_STALE_SECONDS. A pushed state older than this is
-- dropped and the label falls back to the local states, because a dead sender
-- must go quiet rather than leave "on task" under someone who stopped an hour
-- ago.
local PUSH_STALE_SECONDS = 90

-- Mirrors partystatus.PRECEDENCE and partystatus.LOCAL_LABELS. First true one
-- wins. `away` sits above `fighting` on purpose: a member the client cannot
-- see is the failure this exists to catch, and a drifted member is nearly
-- always fighting something.
-- `aboutOthers` marks the two rules that are statements about the DISTANCE
-- between a member and the person watching. Neither can be true of the
-- streamed character himself, and UnitInRange("player") is not a question
-- Blizzard_RaidUI ever asks, so asking it would risk labelling every stream's
-- own subject `apart` forever on the strength of an untested return value.
local LOCAL_RULES = {
    { code = "offline",  label = "offline",
      test = function(u) return not UnitIsConnected(u) end },
    { code = "dead",     label = "dead",
      test = function(u) return UnitIsDeadOrGhost(u) end },
    { code = "away",     label = "away", aboutOthers = true,
      test = function(u) return not UnitIsVisible(u) end },
    { code = "fighting", label = "fighting",
      test = function(u) return UnitAffectingCombat(u) end },
    { code = "apart",    label = "apart", aboutOthers = true,
      test = function(u) return not UnitInRange(u) end },
}

-- The palette, which is the one decision this file owns. Held as a table so a
-- code with no colour is a missing entry the suite can find rather than a
-- silent black label on a dark background.
local COLOUR = {
    offline  = { 0.55, 0.55, 0.55 },
    dead     = { 0.90, 0.25, 0.25 },
    away     = { 1.00, 0.45, 0.10 },
    fighting = { 1.00, 0.82, 0.20 },
    apart    = { 0.95, 0.70, 0.35 },
    task     = { 0.45, 0.85, 0.45 },
    travel   = { 0.45, 0.70, 1.00 },
    dungeon  = { 0.75, 0.55, 1.00 },
    stalled  = { 1.00, 0.35, 0.30 },
    idle     = { 0.60, 0.60, 0.60 },
}

-- Clearance from the right edge of a party frame. See the header: the party
-- background panel ends at x=129 and this puts the label at x=134.
local CLEAR_X = 6

-- The name row's vertical middle. The name's baseline box runs from the frame
-- top down to y=-10 and the health bar starts at y=-12.
local NAME_Y = -5

-- How often the local facts are re-read. UnitInRange and UnitIsVisible fire no
-- events, so there is nothing to listen to; twice a second is far below the
-- rate anything here changes and costs nothing measurable.
local TICK = 0.5

-- name -> { code, label, at }. Only ever written by a line that parsed.
local pushed = {}

-- The units this addon draws, in the order the frames appear on screen.
local UNITS = {
    { unit = "player",  anchor = "player" },
    { unit = "party1",  anchor = "PartyMemberFrame1" },
    { unit = "party2",  anchor = "PartyMemberFrame2" },
    { unit = "party3",  anchor = "PartyMemberFrame3" },
    { unit = "party4",  anchor = "PartyMemberFrame4" },
}

local labels = {}

local function makeLabel(spec)
    local parent
    if spec.anchor == "player" then
        parent = PlayerFrame
    else
        parent = _G[spec.anchor]
    end
    if not parent then return nil end

    local frame = CreateFrame("Frame", nil, parent)
    frame:SetWidth(90)
    frame:SetHeight(12)
    -- Never takes a click. The label sits over the 3D world and anything that
    -- swallowed a click there would break targeting for a person playing.
    frame:EnableMouse(false)

    local text = frame:CreateFontString(nil, "OVERLAY")
    -- OUTLINE rather than a shadow: this is read off a video stream, where a
    -- one-pixel shadow is the first thing an encoder throws away.
    text:SetFont(STANDARD_TEXT_FONT or "Fonts\\FRIZQT__.TTF", 10, "OUTLINE")
    text:SetJustifyH("LEFT")

    if spec.anchor == "player" then
        frame:SetPoint("TOPLEFT", parent, "BOTTOMLEFT", 29, -16)
        text:SetPoint("LEFT", frame, "LEFT", 0, 0)
    else
        frame:SetPoint("LEFT", parent, "TOPRIGHT", CLEAR_X, NAME_Y)
        text:SetPoint("LEFT", frame, "LEFT", 0, 0)
    end

    frame.text = text
    return frame
end

-- --- the pushed half ---------------------------------------------------------

local function absorb(message)
    -- One line, split into name/code/label triples. No decision is taken here
    -- and none is available to take: the words arrived already chosen.
    if not message then return false end
    local fields = { strsplit(SEP, message) }
    if fields[1] ~= PREFIX then return false end
    if fields[2] ~= VERSION then
        -- A version this build does not know is treated exactly like silence.
        -- Guessing at an unknown layout is how a label ends up naming the
        -- wrong character.
        return true
    end
    local now = GetTime()
    local i = 3
    while fields[i + 2] ~= nil do
        local name, code, label = fields[i], fields[i + 1], fields[i + 2]
        if name ~= "" and COLOUR[code] then
            pushed[name] = { code = code, label = label, at = now }
        end
        i = i + 3
    end
    return true
end

-- The party-chat route, and its own suppression in one function. mod-overseer
-- can already broadcast a party message today, so that is what works with no
-- server change; the addon-channel route needs four lines in DoChat and is
-- handled by the CHAT_MSG_ADDON branch below. Returning true removes the line
-- from every chat frame, which is what keeps machine traffic off the stream.
local function chatFilter(_, _, message)
    if message and strsub(message, 1, strlen(PREFIX) + 1) == PREFIX .. SEP then
        absorb(message)
        return true
    end
    return false
end

-- --- painting ----------------------------------------------------------------

local function livePush(unit)
    local name = UnitName(unit)
    if not name then return nil end
    local entry = pushed[name]
    if not entry then return nil end
    if (GetTime() - entry.at) > PUSH_STALE_SECONDS then return nil end
    return entry
end

local function paint(spec, frame)
    local unit = spec.unit
    if not UnitExists(unit) then
        frame:Hide()
        return
    end
    for _, rule in ipairs(LOCAL_RULES) do
        if not (rule.aboutOthers and unit == "player") and rule.test(unit) then
            local c = COLOUR[rule.code]
            frame.text:SetText(rule.label)
            frame.text:SetTextColor(c[1], c[2], c[3])
            frame:Show()
            return
        end
    end
    local entry = livePush(unit)
    if entry then
        local c = COLOUR[entry.code]
        frame.text:SetText(entry.label)
        frame.text:SetTextColor(c[1], c[2], c[3])
        frame:Show()
        return
    end
    -- Nothing to say. An empty label rather than a cheerful one: with no
    -- sender running there is no honest way to tell "with the party, working"
    -- from "with the party, being driven by nothing".
    frame.text:SetText("")
    frame:Hide()
end

local function repaint()
    for i, spec in ipairs(UNITS) do
        local frame = labels[i]
        if frame then
            -- pcall: a missing unit API on an unexpected client build must
            -- cost one blank label, not the whole addon and every label after
            -- it. OverseerCam takes the same care with SetCVar.
            pcall(paint, spec, frame)
        end
    end
end

local driver = CreateFrame("Frame")
driver:RegisterEvent("PLAYER_LOGIN")
driver:RegisterEvent("PLAYER_ENTERING_WORLD")
driver:RegisterEvent("PARTY_MEMBERS_CHANGED")
driver:RegisterEvent("CHAT_MSG_ADDON")

driver:SetScript("OnEvent", function(_, event, ...)
    if event == "CHAT_MSG_ADDON" then
        local prefix, message = ...
        if prefix == PREFIX then
            absorb(PREFIX .. SEP .. message)
        end
        return
    end
    if not labels[1] then
        for i, spec in ipairs(UNITS) do
            labels[i] = makeLabel(spec)
        end
        ChatFrame_AddMessageEventFilter("CHAT_MSG_PARTY", chatFilter)
        ChatFrame_AddMessageEventFilter("CHAT_MSG_RAID", chatFilter)
    end
    repaint()
end)

local since = 0
driver:SetScript("OnUpdate", function(_, elapsed)
    since = since + elapsed
    if since < TICK then return end
    since = 0
    repaint()
end)

-- /pstat - say what each label currently reads, and where its answer came
-- from. There is no mouse on an unattended client and no tooltip to hover, so
-- this is how a person checks the addon agrees with the world.
SLASH_PARTYSTATUS1 = "/pstat"
SlashCmdList["PARTYSTATUS"] = function()
    for _, spec in ipairs(UNITS) do
        if UnitExists(spec.unit) then
            local name = UnitName(spec.unit) or spec.unit
            local shown, source = "", "nothing to say"
            for _, rule in ipairs(LOCAL_RULES) do
                if shown == "" and not (rule.aboutOthers and spec.unit == "player")
                        and rule.test(spec.unit) then
                    shown, source = rule.label, "seen here"
                end
            end
            if shown == "" then
                local entry = livePush(spec.unit)
                if entry then shown, source = entry.label, "pushed" end
            end
            DEFAULT_CHAT_FRAME:AddMessage(
                "PartyStatus: " .. name .. " - " .. shown .. " (" .. source .. ")")
        end
    end
end
